### DISCUSSION TITLE
Error 13 while adding library into Arduino IDE

### DISCUSSION THREAD

**Discussion by marosd** [pschatzmann/arduino-audio-tools#1494] [Q&A] [answered]
Hello,

at first, thank you for your work. It looks really great. 

I am trying to add libraries into Arduino IDE (2.3.2). I had no problem to add "ESP-A2DP.zip" but Audio Tools library is not possible to add.

I have this error while sketch -> include library -> add .zip library:

![obrázok](https://github.com/pschatzmann/arduino-audio-tools/assets/10350177/460d7701-73ed-44d8-aca5-4d1e01444b1b)

Could you please help with that?  Thanks a lot.

PS: I just want to use a simple bluetooth / I2S converter with ESP32. 

Maros

**Comment by pschatzmann** [comment, marked as answer]
I am not using Windows: so Google is your friend!
You can also try to install the library with the help of git like indicated in the Readme

**Reply by marosd** [reply]
Thanks.

I am hardware engineer and this "modern" tools are not familiar. I am good in C, but all this github configs, stupid arduino IDEs (impossible to simply select COM port, baudrate for download code in boot mode) are beyond my future interest, its like a pain :( I just want to make a good BT receiver and use it. 

Maybe stupid question but: If I only use your example from A2DP library (bt_music_receiver_arduino_i2s) which is possible to compile it... is it OK? 
EDIT: Ah... its probably only with connection with Arduino HW. I got my own IS2 HW.

You wrote something about ending support in 3.0.0 Arduino, but in my case with 2.3.2... maybe its enough. 

Thanks a lot,
Maros

**Reply by pschatzmann** [reply]
I am sorry, but none of your comments do not make any sense to me: Nobody forces to use you Arduino and there are no mandatory dependencies to Arduino neither from the A2DP library nor from the AudioTools. You can work with IDF if you prefer to.

In the end it depends what IDF version you want to use: With 4 you can still [use old I2S functionality](https://github.com/pschatzmann/ESP32-A2DP/wiki/Legacy-I2S-API) which has been integrated into the A2DP library. With 5 you need to decide yourself if you want to use the AudioTools or just output the data that is provided via the callback function or just use any other I2S library.

But be warned: the new IDF I2S API is C and it is horrible! But maybe in this case it is not too bad because you can hardcode everything.

My current position is, that I don't want to replicate the same i2s functionality into different projects and I have already removed it from my other libraries that had the same issue.

**Reply by marosd** [reply]
Dear Phil,

sorry for misleading lines... I was frustrated to not understand whole thing. I have ESP32 WROOM module with external anthena. I made a custom board for it (I plan to download code to ESP32 via common USB/UART converter). So I left only TTL pins on the board. I2S otputs from module are roated to some external ICs.. I feel, that download code into that will be pain also (I need probably to configure custom board..)

I downloaded Arduino IDE because I use it several years ago, maybe wrong move. And I would like to test BT receiver. I maybe saw it too simple. I am so confused with everything. It seems to me like I "have to" use modules from ESP, Adafruit and so on and just wire it. Any custom made board is pain.

1. With your old A2DP library you said I need IDF 4 (for example), but this IDF is somehow added in Arduino IDE? Because I do not install anything. I am just downloaded IDE, and add your A2DP library like a .zip. 

I am maybe too old to understand but I do not know how all libraries and dependancies are "included" together. I look into yout library and found in "BluetoothA2DPCommon.h" is included "esp_idf_version.h" (for example). So this is the IDF library? And if yes, how is it "included" into Arduino IDE? Via that targed board? 


You are too kind to answer to me and I do not really waste your time. 

All the best,
Maros

**Reply by pschatzmann** [reply]
I see, you don't understand the software architecture: IDF is provided by Espressif to program their chips. The ESP32 Arduino core is just a configured IDF project that has IDF included and provides the Arduino API on top of it. 

Different  Arduino ESP32 core versions come with different IDF versions. This is usually documented in the [release information](https://github.com/espressif/arduino-esp32/releases) of the ESP32 Arduino core project.

The ESP32 specific functionality of my audio libraries build upon the IDF functionality, so they work with IDF, Arduino and PlatformIO:
- The advantage of Arduino is that it is easy to use.
- The advantage of IDF is that you can heavily configure the functionality, but it is quite difficult to use

**Reply by pschatzmann** [reply]
What I2S DAC/Audio Module are you using ?
I started to collect some drivers in my [arduino-audio-driver](https://github.com/pschatzmann/arduino-audio-driver) library, so that they can be used easily in the AudioTools.

**Reply by HobmanX** [reply]
maros.... you need to download the Audio-tools library ..... go to the github  ... code.....download the Zip file....copy it the your arduino sketchbook libraries folder......Unzip it ....run Arduino IDE open your sketch....#include AudioTools.h and Arduino will pick it up......??    .............  or git clone it.

**Reply by marosd** [reply]
Dear gentlemen,

during waiting I installed Espresiff IDE based on Eclipse. It MAYBE works but yes, Phil, you were right. Its not so clean, but at least I can understand dependacies and so on. A lot of troubles to compile it. I work around it for a while and I will return back to try Arduino IDE and you library just for simplicity. I will try HobmanX solution! Thanks a lot.

But more important problem is, that I bought ESP32-C6 WROOM and it not support A2DP, because it do not have classic BT. I found information only on some forum. Question is WHY is not marked in ESP pages.

Very strange is that is IMPOSIBLE to find reference manual for ESP32-C6 or ESP32 WROOM thru their web pages. Or its possible, but I really do not understand how **************** hard is to get some normal information about the usage. I do not care about datasheet, where are described EMC norms, transmit levels, timings ..... I just need to know firstly, what is included and could be configured.

For example I use STM32 and on their web pages is everything clear - datasheets, reference manuals, app notes, tools... Every STM32 has cross table where is marked what is differnet, what is similar.

**Comment by pschatzmann** [comment]
I am not sure but using Google and serching for [Technical Sheet ESP32-C6 ](https://www.google.com/search?q=technical+sheet+ESP32-C6&oq=technical+sheet+ESP32-C6&gs_lcrp=EgZjaHJvbWUqBggAEEUYOzIGCAAQRRg70gEIMTAyN2owajeoAgCwAgA&sourceid=chrome&ie=UTF-8) seems to provide all relevant information

**Reply by marosd** [reply]
I know, but as I wrote, try it directly from their web. I just mentioned it, because for beginner in the ESP platform is pretty hard to start something and lot of partial information. Sorry for spam here.

Maros
